### ISSUE TITLE
[MLv2] Functions for `drill-thru`

### ISSUE THREAD

**Issue by camsaul** [metabase/metabase#30081]
Drill-thru is when you have a table of results and you click a particular cell and it gives you a little context menu like

<img width="212" alt="image" src="https://user-images.githubusercontent.com/1455846/231841928-755f3b51-28a8-4273-923a-ba9588148bf9.png">

and then you click one of those options and it generates a new version of a query with a filter on that column e.g.

<img width="207" alt="image" src="https://user-images.githubusercontent.com/1455846/231842521-66961eb9-c3c5-4d18-a1e3-78024d0185ae.png">

<img width="262" alt="image" src="https://user-images.githubusercontent.com/1455846/231842541-ba879c4b-fd64-47b2-a43c-9450778c1d33.png">

Different columns have different drill thru options, e.g. plain number columns have filter options like these:

<img width="254" alt="image" src="https://user-images.githubusercontent.com/1455846/231842909-6ff53f13-2b11-422c-a5de-25ad062cf496.png">

Also, if you click a column itself, you have additional options like these:

<img width="189" alt="image" src="https://user-images.githubusercontent.com/1455846/231844921-1c76e03a-1090-4c3f-85da-63bba7581822.png">

There's a lot more to drill-thru than just this, but this is the basic idea. See all the stuff in https://github.com/metabase/metabase/tree/decc8b7d5733944e5e11e8822e88f9b6ccdd9a4c/frontend/src/metabase-lib/queries/drills for example. See also https://github.com/metabase/metabase/blob/766febd7003e054ea5fe996843cdac3be9418155/frontend/src/metabase-lib/queries/utils/drilldown.js, one of those rare parts of the frontend codebase that is documented.

Drill thru is conceptually a single operation and thus should be exposed in MLv2 as a single operation, in other words we should not ask the frontend to manually build drill thru behaviors out of several lower-level operations like `orderBy()` and the like; these should be util functions in MLv2 itself.

I think we will need the following functions

* `availableDrillThrus(query, column)` (or `availableDrillThrus(query, column, value)`? Or both?) => return a list of available drill thru operations. Not sure what these should look like, but some sort of **new** opaque object you can get `displayInfo` from is probably appropriate, but I'll leave this as an exercise for the implementer. 
* `displayInfo(query, drillThru)` => get the info you need to render a drill thru as returned by `availableDrillThrus()` in the UI
* `withDrillThru(query, column, drillThru)` and/or `withDrillThru(query, column, drillThru, value)` => return a version of the query with the drill thru as returned by `availableDrillThrus()`

**Comment by camsaul** [comment]
This can and should be broken out into smaller tasks if appropriate

**Comment by kulyk** [comment]
My two cents:

1. I believe we’d most likely want `availableDrillThrus` to accept query + column + value — value is needed for a lot of drills

2.  At first, I was thinking that every drill would have its own MLv2 function (like `Lib.foreignKeyDrill(query, column, value)`), but I think having a way to retrieve all the available drills would be handier and match what we have now

3. I’m curious about what the `display_info` output might look like. E.g. for a foreign key drill we show this “View this {{columnTitle}}’s {{tableTitle}}” (like “View this Product’s Orders”). Do you see it as something MLv2 should generate? That makes sense to me FWIW. On the other hand we show an icon next to almost every drill and I believe that shouldn’t be coming from MLv2. Drills can also be shown in different places: e.g. the sorting drill is used when you click the column header, most of the other drills show up in a popover when you click something on a viz. And even in the popup they can also be shown with a different UI. That makes me think we’d like some sort of a constant ID/key for each drill, so the FE can figure out how to present them

**Comment by camsaul** [comment]
Yeah I think maybe some sort of `type` key in the display info or something is probably okay just this once, so the FE can map it to an appropriate icon. But the title/`displayName` can be calculated as part of the `displayInfo`. Something like 

```json
{
  displayName: "View this Product's Orders",
  type: "drill-thru/x-ray"
}
```

I have no idea about that type key btw, but it could map to a Clojure keyword the same way base/effective types and coercion types do, the difference here is that we'd use `drill-thru/` as the prefix or something. 

That does sort of raise the question of whether that should come back as an opaque keyword and then we expose a bunch of functions like `Lib.isXRayDrillThru(drillThruType)` or something like we were discussing for https://github.com/metabase/metabase/issues/30054, but I want to be careful that we don't turn this library into [Enterprise Fizz Buzz](https://github.com/EnterpriseQualityCoding/FizzBuzzEnterpriseEdition) by overengineering stuff. 

We can bikeshed this a bit when we get around to implementing it

**Comment by camsaul** [comment]
RE `availableDrillThrus()` always taking a value: I think we need a way to differentiate between when you click it on a column header and thus have no value, vs. clicking on a cell where the value is `NULL` -- this is not really a problem in JS when you have `undefined` vs `null`, but in Cljs we just have `nil` so it's not easy to distinguish the two cases. I think we'll probably just want to do both arities; `availableDrillThrus(query, column)` for the column header and `availableDrillThrus(query, column, value)` for an individual cell

**Comment by camsaul** [comment]
Closing as implemented by #31735 

We have follow-on issues to tackle remaining things
